music21/musicxml/m21ToXml.py

```diff
@@ -3076,6 +3083,77 @@ def mainElementsParse(self):
             # Assumes voices are flat...
             self.parseFlatElements(v, backupAfterwards=backupAfterwards)
 
+    def getVoiceNumberOffset(self) -> int:
+        '''
+            Returns the offset that should be added to any voice numbers in this measure,
+            in order to avoid voice numbers in other simultaneous measures within a staff
+            group.  If this measure is not in a staff group, the offset is 0.
+            This is computed once, and cached, for performance.
+        '''
+        if self.voiceNumberOffset is not None:
+            # return the cached computed offset
+            return self.voiceNumberOffset
+
+        def getMaxVoiceNumber(
+                scoreExp: ScoreExporter,
+                partStaff: stream.Part,
+                measure: stream.Measure) -> int:
+            # find the partExporter for partStaff
+            partStaffExp: t.Optional[PartExporter] = None
+            for partx in scoreExp.partExporterList:
+                if partx.stream is partStaff:
+                    partStaffExp = partx
+                    break
+            if partStaffExp is None:
+                return 0
```

Comment:
Realistically this is not so bad, but we could do better than executing a PartStaff search in every single measure.

See the `mergeInstrumentStreamPartStaffAware` model: the search is done once at the part level. Maybe the PartExporter needs a `.joinableGroup` attribute to hold the `StaffGroup` it participates in.